src/utils.cpp

```diff
@@ -735,55 +740,50 @@ std::vector<std::pair<std::string, std::string>> get_cgroup_paths(
-    if (root.first == "cgroup") {
-      root = { std::filesystem::path(root.second).filename().string(),
-               root.second };
-    } else if (root.first == "cgroup2") {
-      root = { "unified", root.second };
-    }
-  }
+  auto types_v1 = roots[0] |
+                  std::views::transform(
+                      [&](auto &root) -> std::pair<std::string, std::string> {
+                        return {
+                          std::filesystem::path(root).filename().string(), root
+                        };
+                      });
+  auto types_v2 = roots[1] |
+                  std::views::transform(
+                      [&](auto &root) -> std::pair<std::string, std::string> {
+                        return { "unified", root };
+                      });
 
   // Filter roots
   bool start_wildcard, end_wildcard;
   auto tokens = get_wildcard_tokens(filter, start_wildcard, end_wildcard);
-  std::vector<std::pair<std::string, std::string>> filtered_roots;
-  std::copy_if(roots.begin(),
-               roots.end(),
-               std::back_inserter(filtered_roots),
-               [&tokens, &start_wildcard, &end_wildcard](auto &pair) {
-                 return wildcard_match(
-                     pair.first, tokens, start_wildcard, end_wildcard);
-               });
+  auto filter_func = std::views::filter([&](auto pair) -> bool {
+    return wildcard_match(pair.first, tokens, start_wildcard, end_wildcard);
+  });
+  auto filtered_v1 = types_v1 | filter_func;
+  auto filtered_v2 = types_v2 | filter_func;
 
   // Get cgroup path for each root
-  std::vector<std::pair<std::string, std::string>> result;
-  std::transform(filtered_roots.begin(),
-                 filtered_roots.end(),
-                 std::back_inserter(result),
-                 [&cgroupid](auto &pair) {
-                   return std::pair<std::string, std::string>{
-                     pair.first,
-                     get_cgroup_path_in_hierarchy(cgroupid, pair.second)
-                   };
-                 });
+  auto get_path_func = std::views::transform(
+      [&](auto pair) -> std::pair<std::string, std::string> {
+        return { pair.first,
+                 get_cgroup_path_in_hierarchy(cgroupid, pair.second) };
+      });
+  auto paths_v1 = filtered_v1 | get_path_func;
+  auto paths_v2 = filtered_v2 | get_path_func;
 
   // Sort paths lexically by name (with the exception of unified, which always
   // comes first)
-  std::sort(result.begin(), result.end(), [](auto &pair1, auto &pair2) {
-    if (pair2.first == "unified")
-      return false;
-    if (pair1.first == "unified")
-      return true;
-    return pair1.first < pair2.first;
-  });
-
-  return result;
+  std::vector<std::pair<std::string, std::string>> sorted_v1(paths_v1.begin(),
+                                                             paths_v1.end());
+  std::sort(sorted_v1.begin(), sorted_v1.end());
+  std::vector<std::pair<std::string, std::string>> sorted(paths_v2.begin(),
```

Comment:
readability nit: I think this line should go under the comment above about having "unified" come first.

Also (and I haven't tried this) do we have to create the temporary `sorted_v1` vector as opposed to sorting paths_v1 in place and then inserting into `sorted`?

Reply:
The views are a bit weird. You can actually sort using the view itself and then concat, but this is a C++25 thing. You *can* solve it if you built your own type erasure bits, but I didn't really want to go down that path. I'm happy to reorganize to make this more readable.